### PR TITLE
Fix gpu column integrals

### LINKED ISSUE #1376: `column_integral_definite!` causes StackOverflow on gpu
`column_integral_definite!` causes a StackOverflow on the gpu due to its function definition:

```julia
function column_integral_definite!(col∫field::Fields.Field, field::Fields.Field)
    Fields.bycolumn(axes(field)) do colidx
        column_integral_definite!(col∫field[colidx], field[colidx])
        nothing
    end
    return nothing
end
```

This is the next issue we're hitting for moisture cases cc @simonbyrne

### PR DESCRIPTION
**Pull request by charleskawczynski**
Closes #1376.

### CODE REVIEW

File: src/Operators/integrals.jl [hunk 1]
```diff
@@ -7,23 +7,68 @@
...
+    nthreads, nblocks = Spaces._configure_threadblock(Ni * Nj * Nh)
+    @cuda threads = nthreads blocks = nblocks column_integral_definite_kernel!(
+        col∫field,
+        field,
+    )
+end
+
+function column_integral_definite_kernel!(
+    col∫field::Fields.ExtrudedFiniteDifferenceField,
+    field::Fields.ExtrudedFiniteDifferenceField,
```
[the remark this answers is not in the document]
Comment (simonbyrne) [reply src/Operators/integrals.jl:32]: Also isn't `col∫field` typically a `SpectralElementField`?

[the remark this answers is not in the document]
Comment (charleskawczynski) [reply src/Operators/integrals.jl:32]: The column field case is caught below, and `col∫field` can be a `SpectralElementField` or a `PointField` (which is caught in the case below too).

[the remark this answers is not in the document]
Comment (simonbyrne) [reply src/Operators/integrals.jl:32]: No I meant that if `field isa ExtrudedFiniteDifferenceField`, won't `col∫field` be a `SpectralElementField`?

[the remark this answers is not in the document]
Comment (simonbyrne) [reply src/Operators/integrals.jl:32]: Also, this should work fine for column fields as well, avoiding the need for the extra method

[the remark this answers is not in the document]
Comment (charleskawczynski) [reply src/Operators/integrals.jl:32]: Oh, yeah, good catch!

[the remark this answers is not in the document]
Comment (charleskawczynski) [reply src/Operators/integrals.jl]: I think I fixed the dispatch pattern

**Comment by charleskawczynski** [comment]
Now, locally, I'm getting:

```
Definite column integrals bycolumn: Error During Test at dev\ClimaCore.jl\test\Fields\field.jl:693
  Got exception outside of a @test
  InvalidIRError: compiling MethodInstance for ClimaCore.Operators.column_integral_definite_kernel!(::ClimaCore.Fields.Field{IJFH{Float64, 4, SubArray{Float64, 4, CUDA.CuDeviceArray{Float64, 5, 1}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}}, true}}, ClimaCore.Spaces.SpectralElementSpace2D{Nothing, ClimaCore.Spaces.Quadratures.GLL{4}, ClimaCore.Geometry.SphericalGlobalGeometry{Float64}, IJFH{ClimaCore.Geometry.LocalGeometry{(1, 2, 3), ClimaCore.Geometry.LatLongZPoint{Float64}, Float64, SMatrix{3, 3, Float64, 9}}, 4, SubArray{Float64, 4, CUDA.CuDeviceArray{Float64, 5, 1}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}}, true}}, IJFH{Float64, 4, CUDA.CuDeviceArray{Float64, 4, 1}}, ClimaCore.DataLayouts.IFH{ClimaCore.Geometry.SurfaceGeometry{Float64, ClimaCore.Geometry.UVVector{Float64}}, 4, CUDA.CuDeviceArray{Float64, 3, 1}}, NamedTuple{(), Tuple{}}}}, ::ClimaCore.Fields.Field{ClimaCore.DataLayouts.VIJFH{Float64, 4, SubArray{Float64, 5, CUDA.CuDeviceArray{Float64, 5, 1}, Tuple{Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, Base.Slice{Base.OneTo{Int64}}, UnitRange{Int64}, Base.Slice{Base.OneTo{Int64}}}, false}}, ClimaCore.Spaces.ExtrudedFiniteDifferenceSpace{ClimaCore.Spaces.CellCenter, ClimaCore.Spaces.SpectralElementSpace2D{Nothing, ClimaCore.Spaces.Quadratures.GLL{4}, ClimaCore.Geometry.SphericalGlobalGeometry{Float64}, IJFH{ClimaCore.Geometry.LocalGeometry{(1, 2), ClimaCore.Geometry.LatLongPoint{Float64}, Float64, SMatrix{2, 2, Float64, 4}}, 4, CUDA.CuDeviceArray{Float64, 4, 1}}, IJFH{Float64, 4, CUDA.CuDeviceArray{Float64, 4, 1}}, ClimaCore.DataLayouts.IFH{ClimaCore.Geometry.SurfaceGeometry{Float64, ClimaCore.Geometry.UVVector{Float64}}, 4, CUDA.CuDeviceArray{Float64, 3, 1}}, NamedTuple{(), Tuple{}}}, ClimaCore.Topologies.DeviceIntervalTopology{NamedTuple{(:bottom, :top), Tuple{Int64, Int64}}}, ClimaCore.Spaces.Flat, ClimaCore.Geometry.SphericalGlobalGeometry{Float64}, ClimaCore.DataLayouts.VIJFH{ClimaCore.Geometry.LocalGeometry{(1, 2, 3), ClimaCore.Geometry.LatLongZPoint{Float64}, Float64, SMatrix{3, 3, Float64, 9}}, 4, CUDA.CuDeviceArray{Float64, 5, 1}}, ClimaCore.DataLayouts.VIJFH{ClimaCore.Geometry.LocalGeometry{(1, 2, 3), ClimaCore.Geometry.LatLongZPoint{Float64}, Float64, SMatrix{3, 3, Float64, 9}}, 4, CUDA.CuDeviceArray{Float64, 5, 1}}}}) resulted in invalid LLVM IR
  Reason: unsupported call through a literal pointer (call to memcmp)
  Stacktrace:
   [1] _memcmp
     @ .\strings\string.jl:113
   [2] ==
     @ .\strings\string.jl:125
   [3] multiple call sites
     @ unknown:0
  Reason: unsupported dynamic function invocation (call to ClimaComms.MPICommsContext)
  Stacktrace:
   [1] MPICommsContext
     @ C:\Users\kawcz\.julia\packages\ClimaComms\jYrr9\src\mpi.jl:16
   [2] context
     @ C:\Users\kawcz\.julia\packages\ClimaComms\jYrr9\src\context.jl:28
   [3] context
     @ dev\ClimaCore.jl\src\Fields\Fields.jl:43
   [4] column
     @ dev\ClimaCore.jl\src\Spaces\spectralelement.jl:655
   [5] column
     @ dev\ClimaCore.jl\src\Fields\Fields.jl:161
   [6] column_integral_definite_kernel!
     @ dev\ClimaCore.jl\src\Operators\integrals.jl:40
  Reason: unsupported call through a literal pointer (call to ijl_alloc_array_1d)
  Stacktrace:
   [1] Array
```

**Comment by charleskawczynski** [comment]
So, I suspect

```julia
function _column_integral_definite!(
    col∫field::Fields.PointField,
    field::Fields.ColumnField,
)
    @inbounds col∫field[] = _column_integral_definite(field)
    return nothing
end

function _column_integral_definite(field::Fields.ColumnField)
    field_data = Fields.field_values(field)
    Δz_data = Spaces.Δz_data(axes(field))
    Nv = Spaces.nlevels(axes(field))
    ∫field = zero(eltype(field))
    @inbounds for j in 1:Nv
        ∫field += field_data[j] * Δz_data[j]
    end
    return ∫field
end
```
is not cuda-capable

**Comment by simonbyrne** [comment]
the error appears to be the call to `column`, specifically this line:
https://github.com/CliMA/ClimaCore.jl/blob/14ae29965a5352ef0ca4342bd32240779e4cd286/src/Spaces/spectralelement.jl#L655
I guess we didn't test the thomas algorithm on the GPU.

**Comment by charleskawczynski** [comment]
> the error appears to be the call to `column`, specifically this line:
> 
> https://github.com/CliMA/ClimaCore.jl/blob/14ae29965a5352ef0ca4342bd32240779e4cd286/src/Spaces/spectralelement.jl#L655
> 
> 
> I guess we didn't test the thomas algorithm on the GPU with MPI

Bummer, ok. What's the fix here? I don't immediately see what is wrong with a call to `ClimaComms.context`. It looks like it just leads to a `getproperty` call

**Comment by charleskawczynski** [comment]
bors try

**Comment by charleskawczynski** [comment]
bors try

**Comment by charleskawczynski** [comment]
@simonbyrne, should we just use `if !(space isa PointSpace)` around `allreduce!(context, ...)` in the reduction operators? (`sum`, `maximum` etc.)?

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
I think this is ready to go.

**Comment by simonbyrne** [comment]
bors try

**Review by simonbyrne — APPROVED** [review]

**Comment by simonbyrne** [comment]
Okay, it looks like it works on ClimaAtmos:
https://buildkite.com/clima/climaatmos-ci/builds/11371#0189797b-ed7f-4290-80fb-06f7529bc1da
so i'll squash and merge.

**Pull request merged**